components/net/resource_thread.rs

```diff
@@ -429,10 +431,116 @@ pub struct CoreResourceManager {
     devtools_chan: Option<Sender<DevtoolsControlMsg>>,
     swmanager_chan: Option<IpcSender<CustomResponseMediator>>,
     filemanager: FileManager,
-    fetch_pool: rayon::ThreadPool,
+    thread_pool: Arc<CoreResourceThreadPool>,
     certificate_path: Option<String>,
 }
 
+/// The state of the thread-pool used by CoreResource.
+enum ThreadPoolState {
+    /// Pool is inactive, cannot spawn additional work,
+    /// and waiting on still active workers to shut-down.
+    Inactive(u32),
+    /// Pool is active, and counting active workers.
+    Active(u32),
+}
+
+// Threadpool used by Fetch and file operations.
+pub struct CoreResourceThreadPool {
+    pool: rayon::ThreadPool,
+    state: Arc<Mutex<ThreadPoolState>>,
+}
+
+impl CoreResourceThreadPool {
+    pub fn new(num_threads: usize) -> CoreResourceThreadPool {
+        let pool = rayon::ThreadPoolBuilder::new()
+            .num_threads(num_threads)
+            .build()
+            .unwrap();
+        let state = Arc::new(Mutex::new(ThreadPoolState::Active(0)));
+        CoreResourceThreadPool { pool: pool, state }
+    }
+
+    // Spawn work on the thread-pool, if still active.
+    //
+    // There is no need to give feedback to the caller,
+    // because if we do not perform work,
+    // it is because the system as a whole is exiting.
+    pub fn spawn<OP>(&self, work: OP)
+    where
+        OP: FnOnce() + Send + 'static,
+    {
+        {
+            let mut state = self.state.lock().unwrap();
+            match &mut *state {
+                ThreadPoolState::Active(num) => {
+                    // Increment number of active workers.
+                    *num += 1;
+                },
+                ThreadPoolState::Inactive(_) => return,
```

Comment:
Probably should print a warning to the console when this happens? This shouldn't happen, right?

Reply:
It can happen, because fetch runs in parallel to the main core resource thread, and could try to read a file in chunks after the main thread is trying to exit.

Reply:
As a side, this is sort of "expected", in the sense that the core resource manager is in the process of exiting, and then fetch would still try to spawn additional work. 

The "the weak cannot be upgraded" sitution noted in the other comment, would be if a worker is running even after the core resource manager has already dropped. 

Do you think we still need to warn here? I would say the return here is perhaps not worth warning about since it's a fairly normal situation.

Reply:
Ah, if it's expected then never mind. I thought it would only happen if something went wrong in the shutdown order.